mm/iob/iob_alloc.c

```diff
@@ -163,36 +172,6 @@ static FAR struct iob_s *iob_allocwait(bool throttled, unsigned int timeout)
            */
 
           iob = iob_alloc_committed();
-          if (iob == NULL)
-            {
-              /* We need release our count so that it is available to
-               * iob_tryalloc(), perhaps allowing another thread to take our
-               * count.  In that event, iob_tryalloc() will fail above and
-               * we will have to wait again.
-               */
-
-              sem->semcount++;
```

Comment:
why skip increasement

Reply:
As soon as an IOB buffer is freed, it becomes available in `iob_alloc_committed` if a task is waiting for a throttled or non-throttled IOB. We eliminated the need to increase the semaphore momentaneously and calling `tryalloc`: IOBs in  `g_iob_committed` do not rely on semaphores values to be allocated. They were already incremented by `nxsem_post` and just acquired by the waiting task by calling `iob_alloc_committed`.

### Example: Waiting for a throttled IOB

#### Before this commit:
- A task is waiting for a throttled IOB buffer but we have non-throttled IOB buffers available (`g_iob_buffer` is positive and `g_throttled_sem` negative).
- An IOB is just freed: it will be put in the free list (`g_iob_freelist`) instead of the committed list.
- The task waiting for the throttled IOB will be awakened by posting to the `g_throttled_sem` (that may become equal to `0`, if only a single thread is waiting).
- We need to temporarily increase the `g_throttled_sem` to be able to alloc the throttled IOB in `iob_tryalloc`.
 
This has some flaws: we may suffer with a throttled/non-throttled mismatch in the SMP case, for instance, and `tryalloc` may fail and we had already incremented the semaphore value (which would not be decremented properly).

#### Current implementation:

We avoid making any "single" manipulations with semaphore values. They are only incremented and decrement together. Considering the same example:
- As soon as the IOB is freed and we have a task waiting for it (either throttled or non-throttled), the IOB is put in the committed list.
- Both semaphores values are incremented by `nxsem_post`.
- If committed, at least one task will be awakened. It will acquire the IOB from the committed list and decrement it in `iob_alloc_committed`. The semaphore counter means the amount of free or committed semaphores.

Reply:
@xiaoxiang781216, did this explanation make sense for you? Anything else I can contribute to it?

Reply:
Sorry, for later response, I will review today.

Reply:
@xiaoxiang781216 , were you able to take a look?

Reply:
@wengzhe , can you please take a look too?